File: packages/java/src/main/java/io/opensms/Suppressions.java
package io.opensms;

import io.opensms.models.Page;
import io.opensms.models.Suppression;
import io.opensms.models.SuppressionImportResult;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** The do-not-send list. Accessed as {@code client.suppressions()}. */
public final class Suppressions {

    private final ApiTransport t;

    Suppressions(ApiTransport t) {
        this.t = t;
    }

    /** @return the first page of suppressions. */
    public Page<Suppression> list() {
        return list(new ListParams());
    }

    /**
     * @param params paging.
     * @return one page.
     */
    public Page<Suppression> list(ListParams params) {
        return t.get("/v1/compliance/suppressions", params.toQuery(), ApiTransport.pageOf(Suppression.class));
    }

    /**
     * Suppress one number. Not retried automatically.
     *
     * @param e164   phone number.
     * @param reason {@code stop_keyword}, {@code manual}, {@code complaint} or {@code invalid_number}.
     * @return the suppression.
     */
    public Suppression create(String e164, String reason) {
        return t.json("POST", "/v1/compliance/suppressions", new SuppressionInput(e164, reason).toWire(), null,
                ApiTransport.type(Suppression.class));
    }

    /**
     * Bulk import ({@code import} in the canonical surface, a reserved word in Java).
     * Not retried automatically.
     *
     * @param items numbers and reasons.
     * @return {@code {created, received}}.
     */
    public SuppressionImportResult importItems(List<SuppressionInput> items) {
        if (items == null || items.isEmpty()) {
            throw new IllegalArgumentException("items is required");
        }
        List<Map<String, Object>> wire = new ArrayList<>();
        for (SuppressionInput i : items) wire.add(i.toWire());
        Map<String, Object> body = Wire.map();
        body.put("items", wire);
        return t.json("POST", "/v1/compliance/suppressions/import", body, null,
                ApiTransport.type(SuppressionImportResult.class));
    }

    /** @param id suppression id. */
    public void delete(long id) {
        t.json("DELETE", "/v1/compliance/suppressions/" + id, null, null, ApiTransport.VOID);
    }
}

File: packages/php/src/Resources/Suppressions.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `suppressions` resource: numbers that must never be messaged.
 * Accessed as `$client->suppressions`.
 */
final class Suppressions
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * @param array<string, mixed> $params `limit`, `cursor`
     */
    public function list(array $params = []): Page
    {
        return Models::page($this->transport->request('GET', '/v1/compliance/suppressions', Models::map($params, Models::PAGE_QUERY, [], true)));
    }

    /**
     * Not retried automatically.
     *
     * @param array{e164: string, reason: string} $params reason: stop_keyword|manual|complaint|invalid_number
     * @return array<string, mixed>
     */
    public function create(array $params): array
    {
        $body = Models::map($params, Models::SUPPRESSION, ['e164', 'reason']);

        return Models::object($this->transport->request('POST', '/v1/compliance/suppressions', [], $body, ['retry' => false]));
    }

    /**
     * Bulk add. Returns `{created, received}`. Not retried automatically.
     *
     * @param list<array{e164: string, reason: string}> $items
     * @return array<string, mixed>
     */
    public function import(array $items): array
    {
        $body = ['items' => array_map(
            static fn (array $item): array => Models::map($item, Models::SUPPRESSION, ['e164', 'reason']),
            array_values($items),
        )];

        return Models::object($this->transport->request('POST', '/v1/compliance/suppressions/import', [], $body, ['retry' => false]));
    }

    public function delete(string|int $id): void
    {
        $this->transport->request('DELETE', '/v1/compliance/suppressions/' . Models::pathParam($id));
    }
}
